### PR TITLE
Support XDP Desktop Portal proxy resolver

### PR DESCRIPTION
**Pull request by janbrummer**
In order to make use of libproxy in a flatpak environment, we need to support XDP proxy resolver. Add support.

@DimStar77 Could you please have a look and review this one?

### CODE REVIEW

File: meson_options.txt
Comment (DimStar77) [inline meson_options.txt:60]: The relevance of this config module is for apps running 'as flatpaks' right? So a 'baremetal' install should not have this config module (but it has to be available to be added into the flatpak)?

Comment (janbrummer) [reply meson_options.txt:60]: Yes, it must be part of the runtime. But there is no way to detect it, and i don't think maintainers are that smart :P As it causes no harm on non flatpak system, i would like to keep it.

File: src/backend/plugins/config-xdp/config-xdp.c
Comment (janbrummer) [inline src/backend/plugins/config-xdp/config-xdp.c:60]: Need to set available to FALSE

File: src/backend/plugins/config-xdp/config-xdp.c
Comment (janbrummer) [inline src/backend/plugins/config-xdp/config-xdp.c:66]: dito

File: src/backend/plugins/config-xdp/config-xdp.c
Comment (janbrummer) [inline src/backend/plugins/config-xdp/config-xdp.c:137]: Leaking memory of g_uri_to_string

Comment (janbrummer) [reply src/backend/plugins/config-xdp/config-xdp.c:137]: Check whether variant needs to be freed as well

**Comment by janbrummer** [comment]
Check possibility of adding a flatpak test

**Comment by janbrummer** [comment]
https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/blob/master/elements/components/libproxy.bst?ref_type=heads 

needs to be changed to enable curl and duktape for automatic proxy support.

File: src/backend/plugins/config-xdp/config-xdp.c
Comment (janbrummer) [reply src/backend/plugins/config-xdp/config-xdp.c]: Fixed.

File: src/backend/plugins/config-xdp/config-xdp.c
Comment (janbrummer) [reply src/backend/plugins/config-xdp/config-xdp.c:66]: Fixed.

File: src/backend/plugins/config-xdp/config-xdp.c
Comment (janbrummer) [reply src/backend/plugins/config-xdp/config-xdp.c:60]: Fixed.

File: meson_options.txt
Comment (janbrummer) [reply meson_options.txt:60]: Hei, that was too easy... i'll switch it :)

**Comment by janbrummer** [comment]
@mcatanzaro In case you have some time, could you please have a look at this PR? Works fine in my test flatpak app, but maybe i missed something important....

Next step would be to adjust freedesktop runtime

**Comment by janbrummer** [comment]
> https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/blob/master/elements/components/libproxy.bst?ref_type=heads
> 
> needs to be changed to enable curl and duktape for automatic proxy support.

Recommended options:
```

                "-Dconfig-env=false",
                "-Dconfig-kde=false",
                "-Dconfig-gnome=false",
                "-Dconfig-sysconfig=false",
                "-Dconfig-xdp=true",
                "-Dintrospection=false",
                "-Drelease=true",
                "-Ddocs=false"
```

**Comment by mcatanzaro** [comment]
The freedesktop runtime is not only used by flatpaks. It's also used by host systems, like GNOME OS, and also for images for embedded systems. So a "disable everything" approach doesn't make sense. We should decide which backend to use at runtime, not compile time.

File: meson_options.txt
Comment (mcatanzaro) [inline meson_options.txt:60]: Why false? Nobody will ever enable it. :)

File: src/backend/plugins/config-xdp/config-xdp.c
Comment (mcatanzaro) [inline src/backend/plugins/config-xdp/config-xdp.c:150]: The API contract here is that if `var` is NULL, then `error` is set, so this is unreachable.

**Review by mcatanzaro — APPROVED** [review]
I think you should enable it by default.

**Comment by mcatanzaro** [comment]
Hm, thorny question: what priority should it have, relative to the other settings providers? If this is _not_ the highest-priority provider, then the other providers will get used instead and possibly return bad results because libproxy doesn't realize it's not running on the host system, right? But if this _is_ the highest-priority provider, then all the other non-Windows providers are suddenly no longer getting tested anymore, because almost everyone will have xdg-desktop-portal installed on Linux. And libproxy will probably be slower, too, since it has to rely on sync D-Bus calls.

Hm, another possible problem: the proxy resolver portal could itself be using libproxy, and in practice it will whenever running outside GNOME. So if all backends are always enabled at build time, then this will turn into an infinite loop where libproxy and the portal both try to look up proxy settings from each other. OK, proposed solution: libproxy should check if it's running under flatpak or snap, and only use the portal if so. That should work, right?

File: meson_options.txt
Comment (janbrummer) [reply meson_options.txt:60]: Because we thought it is only used in flatpak runtime and that there is already a manual configuration. Never thought about GNOME OS.

**Comment by janbrummer** [comment]
> Hm, thorny question: what priority should it have, relative to the other settings providers? If this is _not_ the highest-priority provider, then the other providers will get used instead and possibly return bad results because libproxy doesn't realize it's not running on the host system, right? But if this _is_ the highest-priority provider, then all the other non-Windows providers are suddenly no longer getting tested anymore, because almost everyone will have xdg-desktop-portal installed on Linux. And libproxy will probably be slower, too, since it has to rely on sync D-Bus calls.
> 
> Hm, another possible problem: the proxy resolver portal could itself be using libproxy, and in practice it will whenever running outside GNOME. So if all backends are always enabled at build time, then this will turn into an infinite loop where libproxy and the portal both try to look up proxy settings from each other. OK, proposed solution: libproxy should check if it's running under flatpak or snap, and only use the portal if so. That should work, right?

At the moment every plugin is checked, but config-env does have an higher priority. In case no configuration is found, the plugin returns nothing at all. Only iff all plugins returned nothing, px manager will add a direct rule as a global return value. 

The config-xdp plugin already checks whether it is running in a flatpak environment and disables itself in a non-flatpak environment. All the other plugins are still working and that might also be legit as flatpak permission can transfer e.g. access to sysconfig or dconf or kde configuration files..

So that means that the order in a flatpak environment would be the same as in a non flatpak environemnt:

  - config-env
  - all the rest

Sounds fine to me, right?

**Comment by janbrummer** [comment]
> Recommended options:
> 
> ```
>                 "-Dintrospection=false",
>                 "-Drelease=true",
>                 "-Ddocs=false"
> ```

File: meson_options.txt
Comment (janbrummer) [reply meson_options.txt:60]: Changed it back to true.

File: src/backend/plugins/config-xdp/config-xdp.c
Comment (janbrummer) [reply src/backend/plugins/config-xdp/config-xdp.c:150]: Missed it, thanks. Updated.

**Comment by mcatanzaro** [comment]
Yeah OK, sounds good, no problem then. Except you'll want to use the portal under snap too, not just flatpak.

**Comment by janbrummer** [comment]
https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues/1717

**Pull request merged**
